File: src/scripts/time_evaluation.py
import argparse
import os
import time
from datetime import datetime
import pandas as pd
import torch
from sklearn.cluster import KMeans
from sklearn.decomposition import NMF
from sklearn.feature_selection import VarianceThreshold
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler, MinMaxScaler, FunctionTransformer
from mvlearn.decomposition import AJIVE, GroupPCA
from mvlearn.cluster import MultiviewSpectralClustering, MultiviewCoRegSpectralClustering
from imvc.cluster import OSLFIMVC, DAIMC, EEIMVC, LFIMVC, MKKMIK, MSNE, SIMCADC, PIMVC, IMSR, OMVC, OPIMC, SUMO, NEMO, \
    IMSCAGL
from imvc.cluster.monet import MONET
from imvc.decomposition import DFMF, MOFA, DeepMF, jNMF
from imvc.preprocessing import MultiViewTransformer, NormalizerNaN, ConcatenateViews

from src.models import Model
from settings import TIME_RESULTS_PATH, TIME_LOGS_PATH, TIME_ERRORS_PATH, RANDOM_STATE, DATASET_TABLE_PATH
from src.commons import CommonOperations

datasets = pd.read_csv(DATASET_TABLE_PATH)["dataset"].to_list()
if "nutrimouse" in datasets:
    pos = datasets.index('nutrimouse')
    datasets[pos:pos + 1] = ('nutrimouse_genotype', 'nutrimouse_diet')

algorithms = {
    "Concat": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                  ConcatenateViews(),
                                    StandardScaler().set_output(transform='pandas'),
                                    KMeans()), "params": {}},
    "NMF": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                  ConcatenateViews(),
                                  MinMaxScaler().set_output(transform='pandas'),
                                  NMF().set_output(transform='pandas'), StandardScaler(), KMeans()), "params": {}},
    "MVSC": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                  MultiViewTransformer(StandardScaler().set_output(transform= "pandas")),
                                                  MultiviewSpectralClustering()),
                             "params": {}},
    "MVCRSC": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                  MultiViewTransformer(StandardScaler().set_output(transform= "pandas")),
                                                       MultiviewCoRegSpectralClustering()),
                                  "params": {}},
    "GPCA": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                  MultiViewTransformer(StandardScaler()), GroupPCA(), StandardScaler(), KMeans()),
                 "params": {}},
    "AJIVE": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                  MultiViewTransformer(StandardScaler()), AJIVE(),
                                   MultiViewTransformer(FunctionTransformer(pd.DataFrame)), ConcatenateViews(),
                                   StandardScaler(), KMeans()),
              "params": {}},
    "SNF": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                  MultiViewTransformer(StandardScaler().set_output(transform= "pandas"))), "params": {}},
    "DAIMC": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                   MultiViewTransformer(NormalizerNaN().set_output(transform="pandas")),
                                   DAIMC()), "params": {}},
    "EEIMVC": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                    MultiViewTransformer(StandardScaler().set_output(transform="pandas")),
                                    EEIMVC()), "params": {}},
    "IMSCAGL": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                  MultiViewTransformer(NormalizerNaN().set_output(transform="pandas")),
                                   IMSCAGL()), "params": {}},
    "IMSR": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                  MultiViewTransformer(NormalizerNaN().set_output(transform="pandas")),
                                   IMSR()), "params": {}},
    "LFIMVC": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                    MultiViewTransformer(StandardScaler().set_output(transform="pandas")),
                                    LFIMVC()), "params": {}},
    "MKKMIK": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                    MultiViewTransformer(StandardScaler().set_output(transform="pandas")),
                                    MKKMIK()), "params": {}},
    "MONET": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                   MultiViewTransformer(StandardScaler().set_output(transform="pandas")),
                                   MONET()), "params": {}},
    "MSNE": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                  MultiViewTransformer(StandardScaler().set_output(transform="pandas")),
                                  MSNE()), "params": {}},
    "OMVC": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                  MultiViewTransformer(StandardScaler().set_output(transform="pandas")),
                                  OMVC()), "params": {}},
    "OPIMC": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                   MultiViewTransformer(NormalizerNaN().set_output(transform="pandas")),
                                     OPIMC()), "params": {}},
    "OSLFIMVC": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                      MultiViewTransformer(StandardScaler().set_output(transform="pandas")),
                                      OSLFIMVC()), "params": {}},
    "PIMVC": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                   MultiViewTransformer(NormalizerNaN().set_output(transform="pandas")),
                                   PIMVC()), "params": {}},
    "SIMCADC": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                     MultiViewTransformer(NormalizerNaN().set_output(transform="pandas")),
                                     SIMCADC()), "params": {}},
    "SUMO": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                  MultiViewTransformer(StandardScaler().set_output(transform="pandas")),
                                  SUMO()), "params": {}},
    "DeepMF": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                    ConcatenateViews(), StandardScaler(),
                                    FunctionTransformer(lambda x: torch.from_numpy(x).float().cuda().t()),
                                    DeepMF(), FunctionTransformer(lambda x: x.cpu().detach().numpy()),
                                    StandardScaler(), KMeans()),
                 "params": {}},
    "DFMF": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                  MultiViewTransformer(StandardScaler().set_output(transform="pandas")),
                                  DFMF().set_output(transform="pandas"),
                                  StandardScaler().set_output(transform="pandas"), KMeans()),
             "params": {}},
    "MOFA": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                  MultiViewTransformer(StandardScaler().set_output(transform="pandas")),
                                  MOFA().set_output(transform="pandas"),
                                  ConcatenateViews(), StandardScaler().set_output(transform="pandas"), KMeans()),
             "params": {}},
    "jNMF": {"alg": make_pipeline(MultiViewTransformer(VarianceThreshold().set_output(transform="pandas")),
                                  MultiViewTransformer(MinMaxScaler().set_output(transform="pandas")),
                                  jNMF().set_output(transform="pandas"),
                                  StandardScaler().set_output(transform="pandas"), KMeans()),
             "params": {}},
}

# args = lambda: None
# args.continue_benchmarking, args.n_jobs, args.save_results = True, 2, False
parser = argparse.ArgumentParser()
parser.add_argument('-save_results', default= False, action='store_true')
args = parser.parse_args()
if not args.save_results:
    TIME_RESULTS_PATH = os.path.join("test", "time_evaluation.csv")
    TIME_LOGS_PATH = os.path.join("test", "time_logs.txt")
    TIME_ERRORS_PATH = os.path.join("test", "time_errors.txt")

if os.path.exists(TIME_RESULTS_PATH):
    results = pd.read_csv(TIME_RESULTS_PATH)
else:
    results = pd.DataFrame(algorithms.keys(), columns=["algorithm"]).merge(pd.DataFrame(datasets, columns=["dataset"]),
                                                                           how='cross')
    results["time"] = -1
    results["finished"] = False
    results["completed"] = False
    results["comments"] = "{}"

    os.remove(TIME_LOGS_PATH) if os.path.exists(TIME_LOGS_PATH) else None
    os.remove(TIME_ERRORS_PATH) if os.path.exists(TIME_ERRORS_PATH) else None
    open(TIME_LOGS_PATH, 'w').close()
    open(TIME_ERRORS_PATH, 'w').close()

for dataset_name in datasets:
    Xs, y, n_clusters = CommonOperations.load_Xs(dataset_name=dataset_name)

    for idx, row in results[(~results["finished"]) & (results["dataset"] == dataset_name)].iterrows():
        alg_name = row["algorithm"]
        if alg_name in ["IntNMF", "COCA", "jNMF", "NEMO"]:
            continue
        alg = algorithms[alg_name]
        with open(TIME_LOGS_PATH, "a") as f:
            f.write(f'\n {dataset_name} \t {alg_name} \t {datetime.now()}')

        start_time = time.perf_counter()
        try:
            clusters, _ = Model(alg_name=alg_name, alg=alg).method(train_Xs=Xs, n_clusters=n_clusters,
                                                                   random_state=RANDOM_STATE, run_n=0)
            elapsed_time = time.perf_counter() - start_time
            errors_dict = {}
            completed = True
        except Exception as exception:
            elapsed_time = time.perf_counter() - start_time
            exception_name = type(exception).__name__
            errors_dict = {f"{exception_name}": exception}
            with open(TIME_ERRORS_PATH, "a") as f:
                f.write(f'\n {dataset_name} \t {alg_name} \t {exception_name}: {exception} \t {datetime.now()}')
            completed = False

        results.loc[idx, ["time", "finished", "completed", "comments"]] = [elapsed_time, True, completed, errors_dict]
        results.to_csv(TIME_RESULTS_PATH, index=False)

print("Completed successfully!")
with open(TIME_LOGS_PATH, "a") as f:
    f.write(f'\n Completed successfully \t {datetime.now()}')

File: src/scripts/time_evaluation.R
utils_path = file.path("src", "utils", "utils.R")
source(utils_path)


datasets = rownames(read.csv(DATASET_TABLE_PATH, row.names = 1))

if ("nutrimouse" %in% datasets) {
  pos <- which(datasets == "nutrimouse")
  datasets <- c(datasets[1:(pos-1)], "nutrimouse_genotype", "nutrimouse_diet", datasets[(pos+1):length(datasets)])
}

algorithms = c("IntNMF", "COCA", "jNMF", "NEMO")

results = read.csv(TIME_RESULTS_PATH, stringsAsFactors= FALSE)
convert_to_logical <- function(column) {
  if (is.character(column) && all(column %in% c("True", "False"))) {
    return(as.logical(ifelse(column == "True", TRUE, FALSE)))
  } else {
    return(column)
  }
}
results[] <- lapply(results, convert_to_logical)

for (dataset_name in datasets) {
  names <- strsplit(dataset_name, "_")[[1]]
  if ("simulated" %in% names) {
    names <- paste(names, collapse = "_")
  }
  x_name <- names[1]
  y_name <- ifelse(length(names) > 1, names[2], "X0")

  unfinished_results <- results[results[["dataset"]] == dataset_name,]
  unfinished_results <- unfinished_results[!unfinished_results[["finished"]],]

  if(nrow(unfinished_results) == 0){
        next
    }
  
  data = load_dataset(dataset_name=x_name, return_y=T, shuffle=TRUE, seed= RANDOM_STATE)
  Xs <- data[["X"]]
  y <- data[["y"]]
  y <- y[,y_name]
  n_clusters <- length(unique(y))

  for (i in rownames(unfinished_results)) {
    row <- results[i,]
    if(row[["finished"]]){
        next
    }
    alg_name <- row[["algorithm"]]
    if(!(alg_name %in% algorithms)){
        next
    }
    cat(paste("\n", dataset_name, alg_name, Sys.time(), sep = "\t"),
        file=TIME_LOGS_PATH,append=TRUE)

    start_time <- Sys.time()
    clustering <- tryCatch({
      if (alg_name == "IntNMF") {
        train_Xs = lapply(Xs, function(x){x[,apply(x, 2, var, na.rm=TRUE) != 0]})
        train_Xs = lapply(train_Xs, normalize)
        train_Xs = lapply(train_Xs, as.matrix)
        set.seed(RANDOM_STATE)
        clusters <- IntNMF::nmf.mnnals(dat=train_Xs, k=n_clusters, seed=RANDOM_STATE)$clusters
      } else if (alg_name == "COCA") {
        train_Xs = lapply(Xs, function(x){x[,apply(x, 2, var, na.rm=TRUE) != 0]})
        train_Xs = lapply(train_Xs, scale)
        set.seed(RANDOM_STATE)
        clusters <- coca::buildMOC(train_Xs, M = length(train_Xs), K = n_clusters)$moc
        clusters <- coca::coca(clusters, K = n_clusters)$clusterLabels
      } else if (alg_name == "jNMF") {
        train_Xs = lapply(Xs, function(x){x[,apply(x, 2, var, na.rm=TRUE) != 0]})
        train_Xs = lapply(train_Xs, normalize)
        train_Xs = lapply(train_Xs, as.matrix)
        set.seed(RANDOM_STATE)
        clusters <- nnTensor::jNMF(X= train_Xs, J = n_clusters)$W
        set.seed(RANDOM_STATE)
        clusters <- kmeans(clusters, n_clusters)$cluster
      } else if (alg_name == "NEMO") {
        train_Xs = lapply(Xs, function(x){x[,apply(x, 2, var, na.rm=TRUE) != 0]})
        train_Xs = lapply(train_Xs, scale)
        train_Xs = lapply(train_Xs, t)
        set.seed(RANDOM_STATE)
        clusters <- nemo.clustering(train_Xs, num.clusters= n_clusters)
      }
      stopifnot(length(y) == length(clusters))
      err <- "{}"
      completed <- TRUE
      clustering <- list(err, completed)
    }, error = function(e) {
      cat(paste("\n", dataset_name, alg_name, e, Sys.time(), sep = "\t"),
          file=TIME_ERRORS_PATH,append=TRUE)
      err <- e
      completed <- FALSE
      return(list(err, completed))
    })
    elapsed_time <- as.numeric(difftime(Sys.time(), start_time, units = "secs"))
    results[i, "finished"] = TRUE
    results[i, "completed"] = clustering[[2]]
    results[i, "time"] = elapsed_time
    results[i, "comments"] = as.character(clustering[[1]])

    write.csv(results, file = TIME_RESULTS_PATH, row.names = F)
  }
}
cat("Completed successfully!")
cat("Completed successfully", file=TIME_LOGS_PATH,append=TRUE)
